lwm2m: Check CoAP header offset

mlasch opened the pull request:
There is a possibility where the Registration message is silently dropped when the header, as well as, the payload get too large. This primarily affects applications using long endpoint names and large messages or block transfer.

The LwM2M client has a Kconfig option `CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE` to configure the maximum CoAP header length. The header length defaults to 48 bytes in a range from 24 to 128. For Registration messages, the header size is dynamic and specific to an application. The endpoint name max length configured with `CONFIG_RD_CLIENT_ENDPOINT_NAME_MAX_LENGTH` defaults to 33 (32 characters and a zero termination byte). This may cause the header ending up larger than configured.

When constructing a header for a LwM2M message, the final size is not checked. If the header is larger than the configured maximum size, it is writing into the area meant for the payload in the network packet buffer. In most cases this is not an issue, because the Registration message is rather small. However, if the Registration message would exceed the buffer reserved for the whole packet (determined by the block size), the packet can't be assembled (-ENOMEM). This prevents the packet to be sent, as well as using a block transfer for even larger messages. In addition, since the result of `async_message_send` is not checked, the Registration message would be silently dropped without warning or error to the user.

The issue can be reproduced by extending the endpoint name and as a result increase the header size beyond the limit. The registration message is now lost, if the payload would be too larger that the space left in the buffer. Like when having a large number of objects registered with the client.

```
$ west build -b qemu_x86/atom samples/net/lwm2m_client -- -DCONFIG_LWM2M_APP_ID="\"AAAABBBBCCCCDDDDEEEEFFFFGGGGHHHHIIII\""
$ west build -t run
# all goes well even though the header is too large
```

A Bootstrap message would also exceed the header boundary, but since it has no payload it would no be dropped that easily.

This change adds a boundary check for every constructed CoAP packet, in order to ensure the headers always stay withing the configured limit. A check is also introduced for the result of `async_message_send` to propagate and log an error. This  prevents issues with future modifications to the header which might have effects on the size independent of `CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE`.

With the changes the output would look like this

```
[00:00:00.510,000] <inf> net_lwm2m_rd_client: RD Client started with endpoint 'AAAABBBBCCCCDDDDEEEEFFFFGGGGHHHH' with client lifetime 30 using server object 0
[00:00:00.510,000] <dbg> net_lwm2m_message_handling: lwm2m_parse_peerinfo: Parse url: coap://192.0.2.2:5683
[00:00:00.510,000] <inf> net_lwm2m_engine: Connected, sock id 0
[00:00:00.510,000] <dbg> net_lwm2m_registry: lwm2m_engine_get: path:1/0/1/0, level 3, buf:0x15b6c0, buflen:4
[00:00:00.510,000] <err> net_lwm2m_message_handling: CoAP header size exceeded, increase the value of CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE to at least 71
[00:00:00.510,000] <err> net_lwm2m_rd_client: error -12 when sending registration message
[00:00:00.510,000] <err> net_lwm2m_rd_client: Registration err: -12
[00:00:00.510,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 18
[00:00:00.510,000] <err> net_lwm2m_rd_client: sm_do_network_error, retries 0
```

clearly showing the issue. Same for Bootstrap

```
[00:00:00.110,000] <inf> net_lwm2m_rd_client: Start LWM2M Client: AAAABBBBCCCCDDDDEEEEFFFFGGGGHHHH
[00:00:00.510,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 1
[00:00:00.510,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 2
[00:00:00.510,000] <dbg> net_lwm2m_registry: lwm2m_engine_get: path:0/0/1/0, level 3, buf:0x15bb8f, buflen:1
[00:00:00.510,000] <inf> net_lwm2m_rd_client: Bootstrap started with endpoint 'AAAABBBBCCCCDDDDEEEEFFFFGGGGHHHH' using security object 0
[00:00:00.510,000] <dbg> net_lwm2m_message_handling: lwm2m_parse_peerinfo: Parse url: coap://192.0.2.2:5683
[00:00:00.510,000] <inf> net_lwm2m_engine: Connected, sock id 0
[00:00:00.510,000] <err> net_lwm2m_message_handling: CoAP header size exceeded, increase the value of CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE to at least 52
[00:00:00.510,000] <err> net_lwm2m_rd_client: Bootstrap registration err: -12
[00:00:00.510,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 18
[00:00:00.510,000] <err> net_lwm2m_rd_client: sm_do_network_error, retries 0
```

mlasch committed:
net: lwm2m: Check result when sending messages

`lwm2m_send_message_async` returns critial errors which are not handled in
many cases. The function can return -ENOMEM, -EPERM or other error codes
from functions called.

Signed-off-by: Marc Lasch <marc.lasch@husqvarnagroup.com>

mlasch committed:
net: lwm2m: Check the CoAP header size

Check the CoAP header does not exceed the configured limits after all CoAP
options were addded.

Signed-off-by: Marc Lasch <marc.lasch@husqvarnagroup.com>

mlasch commented:
Some tests fail because they reuse the packet buffer and append CoAP option bytes somewhere in the middle of the packet. The offset of those options is now boundary checked to ensure they stay within the area of `CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE`.
It looks like I have to adjust the tests such, that the packet buffer is reset before reusing it.

mlasch committed:
tests: net: lib: lwm2m: Reset message buffer before reuse

To ensure that CoAP options fit in the configured memory, the message
buffer has to be reset after every test.

Signed-off-by: Marc Lasch <marc.lasch@husqvarnagroup.com>

mlasch committed:
tests: net: lib: lwm2m: Create boundary check stub function

Create a stub function for `lwm2m_check_header_boundary` for fff.

Signed-off-by: Marc Lasch <marc.lasch@husqvarnagroup.com>

SeppoTakalo commented:
At a quick look, I don't really understand why do we even need the `LWM2M_ENGINE_MESSAGE_HEADER_SIZE` Kconfig option.

Why do we need to check it, instead of just ensuring that message sizes are large enough?

How about if we just remove it? I cannot spot any location, where we actually need to know the maximum header size, or where the header size is checked.

mlasch commented:
> Why do we need to check it, instead of just ensuring that message sizes are large enough?

`CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE` is used in `lwm2m_object.h` to calculate the `MAX_PACKET_SIZE` which is used to create the buffers. We need enough space to build the CoAP header and to accommodate a payload block of `CONFIG_LWM2M_COAP_MAX_MSG_SIZE`. `CONFIG_LWM2M_COAP_MAX_MSG_SIZE` is smaller or equals the `CONFIG_LWM2M_COAP_BLOCK_SIZE`.

```
#define MAX_PACKET_SIZE         (CONFIG_LWM2M_COAP_MAX_MSG_SIZE + \
                                 CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE)
```

I agree, the buffer must be large enough. But "large enough" may depend on the application. My change ensures that the headers fit in the designated area buffer. One can easily set `CONFIG_LWM2M_ENGINE_MESSAGE_HEADER_SIZE=128` which should hold for most applications.

SeppoTakalo commented:
> I agree, the buffer must be large enough. But "large enough" may depend on the application. My change ensures that the headers fit in the designated area buffer.

But why fitting the header is important? Why threat it differently?
If we remove the `HEADER_SIZE` Kconfig completely, it does not affect the functionality of the engine. If the header fits, it fits. If you send too much data, it does not fit.

And to me, it looks like CoAP library already takes care that all the options fit into the packet.

So, I don't really see the point of ensuring that the header specifically fits into the packet. It is enough to check that all data fits into the packet, and I assume the engine already does that.

Is there any real issue that this change fix?

mlasch committed:
net: lwm2m: Clean up block-wise transfer buffer config

Select between `LWM2M_COAP_MAX_MSG_SIZE` and `LWM2M_COAP_BLOCK_SIZE` to
calculate `LWM2M_MAX_PAYLOAD_SIZE`. This depends on whether block-wise
transfer is enabled or not.

Signed-off-by: Marc Lasch <marc.lasch@husqvarnagroup.com>

mlasch commented:
> if the header overflows there, the block transfer might not work at all

In case of a Registration message, because the return value of `async_message_send` was not checked the message was silently dropped.

I just added a commit which should make the use of `CONFIG_LWM2M_COAP_MAX_MSG_SIZE` and `CONFIG_LWM2M_COAP_BLOCK_SIZE` a bit clearer.

I also just realized that if I build for `native_sim/native/64` the registration message header is already 1 byte larger than the default setting.

SeppoTakalo commented:
Can you send the `net: lwm2m: Check result when sending messages` commit as a separate PR.
That is a valid bugfix.

For other changes, I'm still not sure if they fix anything.
I have not seen any proof that there really is an error. It seems like the CoAP packet buffers are just allocated with MAX_SIZE+HEADER_SIZE. But other than that, there is really no need to check where the boundary of the header is.
Or at least, I cannot spot any problems related to that.

If I set the `LWM2M_ENGINE_MESSAGE_HEADER_SIZE` to zero. I can still run the LwM2M client and pass all the interoperability tests. So I fail to see why the whole configuration is relevant, as long as the message size is large enough to fit all headers.

So overall, I'm not in favor of adding checks in the code that do not really do anything useful. Being strict on the header boundary might broke some code that is already running and working, but might have left the HEADER_SIZE to its default value.

mlasch commented:
> Can you send the net: lwm2m: Check result when sending messages commit as a separate PR.
> That is a valid bugfix.

#84867

> I have not seen any proof that there really is an error. It seems like the CoAP packet buffers are just allocated with MAX_SIZE+HEADER_SIZE. But other than that, there is really no need to check where the boundary of the header is.
Or at least, I cannot spot any problems related to that.

There are cases where we run out of buffer, if the header is larger than anticipated (which already happens for the native_sim sample client). This happens in case of block-transfer is enabled, as @rlubos mentioned above. The current test cases however do not cover this issue, hence you can run the tests with `LWM2M_ENGINE_MESSAGE_HEADER_SIZE` set to zero. I try to come up with a test to demonstrate this specific instance.

With the fix from #84867, I can reproduce the issue by enabling the block transfer with a small block size and add some sensor object instances.

```
--- a/samples/net/lwm2m_client/src/lwm2m-client.c
+++ b/samples/net/lwm2m_client/src/lwm2m-client.c
@@ -199,6 +199,10 @@ static int lwm2m_setup(void)
        /* IPSO: Timer object */
        init_timer_object();
 
+       for (int i=1;i<10;i++) {
+               lwm2m_create_object_inst(&LWM2M_OBJ(IPSO_OBJECT_TEMP_SENSOR_ID, i));
+       }
+
        return 0;
 }
```

```
west build -p -b native_sim/native/64 zephyr/samples/net/lwm2m_client -- -DCONFIG_LWM2M_COAP_BLOCK_SIZE=64 -DCONFIG_LWM2M_COAP_BLOCK_TRANSFER=y -DCONFIG_LWM2M_IPSO_TEMP_SENSOR_INSTANCE_COUNT=10
```
```
[00:00:00.510,000] <inf> net_lwm2m_rd_client: RD Client started with endpoint 'native_sim' with client lifetime 30 using server object 0
[00:00:00.510,000] <dbg> net_lwm2m_message_handling: lwm2m_parse_peerinfo: Parse url: coap://192.0.2.2:5683
[00:00:00.510,000] <inf> net_lwm2m_engine: Connected, sock id 0
[00:00:00.510,000] <dbg> net_lwm2m_registry: lwm2m_engine_get: path:1/0/1/0, level 3, buf:0xffff8d01e594, buflen:4
[00:00:00.510,000] <err> net_lwm2m_rd_client: error -12 when sending registration message
[00:00:00.510,000] <err> net_lwm2m_rd_client: Registration err: -12
[00:00:00.510,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 18
[00:00:00.510,000] <err> net_lwm2m_rd_client: sm_do_network_error, retries 0
[00:00:00.510,000] <dbg> net_lwm2m_rd_client: lwm2m_rd_client_service: State: 6
```
The expected behavior would be to send the registration message with a block transfer. There is no warning about the header size exceeding (1 byte) the configured value.
But also with block-transfer disabled, i would assume that if I set `CONFIG_LWM2M_COAP_MAX_MSG_SIZE` to a certain value, I can send `CONFIG_LWM2M_COAP_MAX_MSG_SIZE` bytes and not only `CONFIG_LWM2M_COAP_MAX_MSG_SIZE - X` bytes.

SeppoTakalo commented:
> With the fix from #84867, I can reproduce the issue by enabling the block transfer with a small block size and add some sensor object instances.

This is not a header overflow.

When you set a `COAP_BLOCK_SIZE` the default packet size is going to be the same:
```
config LWM2M_COAP_MAX_MSG_SIZE
	int "LWM2M CoAP maximum message size"
	default LWM2M_COAP_BLOCK_SIZE
	help
	  CoAP message size used by LWM2M. Minimum is the block size used
	  in blockwise transfers.
```

And when you check the code you proposed and run it in debugger, the registration message headers fits the packet, but payload does not:
```
# Registration message before adding payload
cpkt = {
    data = 0x459270 <_k_mem_slab_buf_body_encode_buffer_slab> "H\002\224\337^\346\3029+\347Ṳrd\021(9lwm2m=1.0\r",
    offset = 48,
    max_len = 1024,
    hdr_len = 12 '\f',
    opt_len = 36,
    delta = 15
  },

# After adding the payload (the list of objects)
# this is still on the big encode buffer
  $11 = {
    data = 0x459270 <_k_mem_slab_buf_body_encode_buffer_slab> "H\002\224\337^\346\3029+\347Ṳrd\021(9lwm2m=1.0\r",
    offset = 189,
    max_len = 1024,
    hdr_len = 12 '\f',
    opt_len = 36,
    delta = 15
  }
```

As you can see, the header fits the packet, but payload does not.

**There is few errors here:**
1. Packet size is too small for the registration message.
2. Payload block size is too big, it does not fit into a packet. Those two values cannot be the same, the [calculation does not work right](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_message_handling.c#L382-L384). You cannot fit 64 bytes of payload into a packet size 64. **This is the real bug here.**
3. LwM2M specification does not actually define if registration message can use block-wise transfer. It might work, but it is outside of the spec. If we are going to use it, we probably need to add support for CoAP Request-Tag as well.


The same example above, when I set the Kconfig like this
![image](https://github.com/user-attachments/assets/b0f4749c-d107-4b55-9893-dabe29f636ed)
So there is no extra room for header, but enough room for complete blockwise-payload and not enough room for registration message.

I can see that it works just fine against Leshan
![image](https://github.com/user-attachments/assets/20336d99-004d-48ea-9928-711d4f1dffe9)

So my proposal stays: We should get rid of the whole "extra room for header" calculation.
The issue here seem to be just the calculation of block-size. Or more like a verification that those block sizes and message sizes make sense.

mlasch commented:
The issue was addressed with #84916
